File: waterfall.py
# use python 3

class segment_tree:
  def __init__(self, n):
    self.n = n
    self.max = [0 for _ in range(0, 4 * n + 2)]
    self.delta = [0 for _ in range(0, 4 * n + 2)]

  def prop(self, idx):
    if (self.delta[idx] == 0):
      return

    self.set(2 * idx + 1, self.delta[idx])
    self.set(2 * idx + 2, self.delta[idx])

    self.delta[idx] = 0

  def set(self, idx, value):
    self.delta[idx] = self.max[idx] = value

  def update(self, left, right, val):
    self.updateinner(0, 0, self.n-1, left, right, val)

  def updateinner(self, i, lr, rr, left, right, val):
    if (lr == left and rr == right):
      self.set(i, val)
      return

    self.prop(i)
    mid = (lr + rr) // 2
    if (left <= mid):
      self.updateinner(2 * i + 1, lr, mid, left, min(mid,right), val)
    if (mid < right):
      self.updateinner(2 * i + 2, mid+1, rr, max(mid+1, left), right, val)

    self.max[i] = max(self.max[2 * i + 1], self.max[2 * i + 2])

  def query(self, left, right):
    return self.queryinner(0, 0, self.n-1, left, right)

  def queryinner(self, i, lr, rr, left, right):
    if (lr == left and rr == right):
      return self.max[i]

    self.prop(i)
    mid = (lr + rr) // 2
    ret = -1
    if (left <= mid):
      ret = max(ret, self.queryinner(2 * i + 1, lr, mid, left, min(mid, right)))
    if (mid < right):
      ret = max(ret, self.queryinner(2 * i + 2, mid+1, rr, max(mid+1, left), right))

    return ret

class segment:
  def __init__(self,x,y,s,x1,x2,r1,r2):
    self.x = x
    self.y = y
    self.s = s

    self.x1 = x1
    self.x2 = x2
    self.r1 = r1
    self.r2 = r2

def custom_sort(a):
  return a[1]

cases = int(input())

for tt in range(0,cases):

  n = int(input())

  mp = {}
  before = {}

  a = [
    [int(x) for x in input().split()] for _ in range(0, n)
  ]
  a.sort(key=custom_sort)

  for e in a:
    mp[e[0]] = 0
    mp[e[0] + e[2]] = 0

  aa = []
  ks = []

  for e in a:
    x = e[0]
    y = e[1]
    s = e[2]
    x1 = x
    x2 = x + s
    r1 = mp[x1]
    r2 = mp[x2]

    mp[x1] += 1
    mp[x2] += 1

    ks.extend([x1,x2])

    aa.extend([segment(x,y,s,x1,x2,r1,r2)])

  a = aa

  ks = list(set(ks))
  ks.sort()

  last = 0
  for key in ks:
    before[key] = last
    mp[key] += last
    last = mp[key]

  for e in a:
    e.r1 += before[e.x1]
    e.r2 += before[e.x2]

  rope = segment_tree(last)
  norope = segment_tree(last)

  ans = [[0,0] for _ in range(0, last)]

  i = 0
  j = 0

  while (i < n):
    j = i
    while (j < n and a[i].y == a[j].y):
      j += 1

    all = 1 + norope.query(0, last-1)
    k = i
    while k < j:
      e = a[k]

      below_no_rope = 1+ norope.query(before[e.x1], mp[e.x2]-1)
      below_ye_rope = 1 + rope.query(before[e.x1], mp[e.x2]-1)

      ans[k][0] = below_no_rope
      ans[k][1] = max(below_ye_rope, all)

      k += 1

    k = i
    while k < j:
      e = a[k]

      norope.update(e.r1, e.r2, ans[k][0])
      rope.update(e.r1, e.r2, ans[k][1])

      k += 1

    i = j

  print(max(rope.query(0, last-1), norope.query(0, last-1))) # god ruby is so much better

File: waterfall.java
 // Solution to Waterfall Quest for the UCF HSPT Contest
// Written by: Pauline Johnson

import java.util.*;

public class waterfall {
	public static void main(String[] args) {
		Scanner kb = new Scanner(System.in);
		int cc = kb.nextInt();
		for(int c = 1; c <= cc; c++) {

			int n = kb.nextInt();
			Segment[] platforms = new Segment[n];

			int[] xs = new int[2*n];
			for(int i = 0; i < n; i++) {
				int l = kb.nextInt(), y = kb.nextInt();
				int s = kb.nextInt();
				platforms[i] = new Segment(l, y, s);
				xs[2*i] = l;
				xs[2*i+1] = l + s;
			}
			Arrays.sort(xs);

			// Compress x values.
			HashMap<Integer, Integer> map = new HashMap<>();
			int idx = 0;
			for(int i = 0; i < 2*n; i++) {
				int l = xs[i];
				if(!map.containsKey(l)) {
					map.put(l, idx);
					idx++;
				}
			}

			// Sort segments from bottom to top.
			Arrays.sort(platforms);

			// Create segment trees.
			SegmentTree norope = new SegmentTree(idx);
			SegmentTree rope = new SegmentTree(idx);

			for(int i = 0; i < n; ) {
				ArrayList<Segment> segments = new ArrayList<>();
				segments.add(platforms[i]);
				i++;
				while(i < n && platforms[i].y == platforms[i-1].y) {
					segments.add(platforms[i]);
					i++;
				}

				// Find maximum in segment's range.
				long all = norope.query(0, idx-1);
				for(Segment s : segments) {
					int l = map.get(s.l);
					int r = map.get(s.l+ s.s);
					long below = rope.query(l, r);
					rope.set(l, r, Math.max(all, below)+1);
				}

				// Set segment's range to max + !.
				for(Segment s : segments) {
					int l = map.get(s.l);
					int r = map.get(s.l+ s.s);
					long below = norope.query(l, r);
					norope.set(l, r, below+1);
				}

			}

			// Query all segments.
			long ans = rope.query(0, idx-1);
			System.out.println(ans);
		}
	}

	static class Segment implements Comparable<Segment>{
		int l, y, s;
		Segment(int l, int y, int s) {
			this.l = l;
			this.y = y;
			this.s = s;
		}

		public int compareTo(Segment s) {
			return y - s.y;
		}
	}

	static class SegmentTree {
		long[] val;
		long[] lazy;
		int n;

		public SegmentTree(int n) {
			this.n = n;
			val = new long[4 * n + 1];
			lazy = new long[4 * n + 1];
		}

		void set(int left, int right, long value) {
			set(1, 0, n-1, left, right, value);
		}

		void set(int i, int lr, int rr, int left, int right, long value) {
			if(lr == left && rr == right) {
				lazy[i] = value;
				return;
			}
			if(lazy[i] != 0) {
				lazy[2 * i] = lazy[i];
				lazy[2 * i + 1] = lazy[i];
				val[i] = lazy[i];
				lazy[i] = 0;
			}

			int mid = (lr + rr) / 2;;;;
			if(left <= mid) {
				set(i * 2, lr, mid, left, Math.min(mid, right), value);
			}
			if(mid < right) {
				set(i * 2 +1, mid + 1, rr, Math.max(mid + 1, left), right, value);
			}
			val[i] = Math.max(Math.max(val[i * 2], lazy[i*2]), Math.max(val[i * 2 + 1], lazy[i*2+1]));
		}

		long query(int left, int right) {
			return query(1, 0, n-1, left, right);
		}

		long query(int i, int lr, int rr, int left, int right) {

			if(lr == left && rr ==right)
				return Math.max(lazy[i], val[i]);

			if(lazy[i] != 0) {
				lazy[2 * i] = lazy[i];
				lazy[2 * i + 1] = lazy[i];
				val[i] = lazy[i];
				lazy[i] = 0;
			}

			int mid = (lr + rr) / 2;
			long ret = Long.MIN_VALUE;
			if(left <= mid) {
				long l = query(i * 2, lr, mid, left, Math.min(mid,  right));
				ret = Math.max(ret, l);
			}
			if(mid < right) {
				long r = query(i * 2 + 1, mid +1, rr, Math.max(mid + 1, left), right);
				ret = Math.max(ret, r);
			}
			return ret;
		}
	}
}
